### PR TITLE
Testing: Warehouse with TDD

### PR DESCRIPTION
**Pull request by denisdimitrov**

### CODE REVIEW

**Comment by denisdimitrov** [comment]
I had to re-write the whole task because the last one was not working properly and the products had no names.

File: src/warehouse/ProductWarehouse.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                return product;
+            }
+            return null;
+        }
+        if (products.containsKey(product.getName())) {
+            int updateQuantity = products.get(product.getName()).quantity + product.getQuantity();
+            products.get(product.getName()).setQuantity(updateQuantity);
+        }
+        return null;
+    }
```
Comment (EmilGeorgiev) [inline src/warehouse/ProductWarehouse.java]: Every if statementr should be one test case. also if  statement contains  ```product.quantity > 0 & product.maxQuantity > 0 & product.price > 0 & product.quantity <= product.maxQuantity``` also should be tested. This make test many and code hard to test. try to use less if statements. Also  nested ``if``` make code hard to read

Comment (denisdimitrov) [inline src/warehouse/ProductWarehouse.java]: Okey, I will lower the use of if statements.

**Comment by denisdimitrov** [comment]
I removed most of the if statements and I added comments because the code might be hard to read. I also removed most of the tests because they were unnecessary and re-wrote the sellProduct() method because it was broken.

File: src/warehouse/Products.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Created by clouway on 29.07.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class Products {
+    private String name;
+    public int maxQuantity;
+    public int price;
+    public int quantity;
```
Comment (EmilGeorgiev) [inline src/warehouse/Products.java]: use immutable objects.

```java
public class Products {
     public final String name;
     public final int maxQuantity;
     public final int price;
     public final int quantity;
}
```

File: src/warehouse/Products.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        this.name = name;
+        this.quantity = quantity;
+    }
+
+    /**
+     * Getters.
+     *
+     * @return
+     */
+    public int getPrice() {
```
Comment (EmilGeorgiev) [inline src/warehouse/Products.java]: this geters are unnecessary

File: src/warehouse/Products.java (hunk 2 above)
Comment (denisdimitrov) [inline src/warehouse/Products.java]: I fixed it. I will push the changes after a minute or two.

**Comment by denisdimitrov** [comment]
I made the objects immutable and I also removed the getters and the setter.

**Comment by denisdimitrov** [comment]
All of the test are working now with the new maxQuantity.

File: src/warehouse/ProductWarehouse.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    };
+
+    /**
+     * If there is no product with the same name, it adds a new one in the Map via .put method.
+     * If there is already a product with the same name it refreshes it's quantity and returns it via .get method.
+     *
+     * @param product to be added.
+     * @return product.
+     */
+    public Products addProduct(Products product) {
```
Comment (EmilGeorgiev) [inline src/warehouse/ProductWarehouse.java]: here you have duplication of   ```addProduct(producs)```. Rename to ```add(Products)```

File: src/warehouse/ProductWarehouse.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        return product;
+    }
+
+    /**
+     * If the Key already exists in the Map it refreshes it's quantity and returns it via .get method.
+     *
+     * @param product to be sold.
+     * @return product.
+     */
+    public Products sellProduct(Products product) {
```
Comment (EmilGeorgiev) [inline src/warehouse/ProductWarehouse.java]: .

File: src/warehouse/Products.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package warehouse;
+
+/**
+ * Created by clouway on 29.07.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class Products {
```
Comment (EmilGeorgiev) [inline src/warehouse/Products.java]: rename to ```Product```

File: test/warehousetests/AddProductsTests.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * Created by clouway on 29.07.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class AddProductsTests {
+
+    private ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void addProductTest() throws Exception {
```
Comment (EmilGeorgiev) [inline test/warehousetests/AddProductsTests.java]: remove ```Test``` from name of test

File: test/warehousetests/AddProductsTests.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        Products product = new Products("Apples", 20, 20);
+        assertEquals(product, wh.addProduct(product));
+    }
+
+    @Test
+    public void addTheSameProductTest() throws Exception {
```
Comment (EmilGeorgiev) [inline test/warehousetests/AddProductsTests.java]: addProductTwice

File: test/warehousetests/SellProductsTests.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * Created by clouway on 29.07.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class SellProductsTests {
+
+    private ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void sellProductTest() throws Exception {
```
Comment (EmilGeorgiev) [inline test/warehousetests/SellProductsTests.java]: happyPath

File: test/warehousetests/SellProductsTests.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        Products product1 = new Products("Apples", 2, 10);
+        Products product = new Products("Apples", 2, 5);
+        wh.addProduct(product1);
+        wh.sellProduct(product);
+        assertEquals(product, wh.sellProduct(product));
+    }
+
+    @Test
+    public void sellTheSameProductTest() throws Exception {
```
Comment (EmilGeorgiev) [inline test/warehousetests/SellProductsTests.java]: sellProductTwice

**Comment by denisdimitrov** [comment]
I fixed everything you asked for.

File: test/warehousetests/SortProductsTests.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ * Created by clouway on 29.07.16.
+ *
+ * @author Denis Dimitrov <denis.k.dimitrov@gmail.com>
+ */
+public class SortProductsTests {
+
+    ProductWarehouse wh = new ProductWarehouse(100);
+
+    @Test
+    public void checkFirstIndex() throws Exception {
```
Comment (EmilGeorgiev) [inline test/warehousetests/SortProductsTests.java]: this test is unnecessary. Actually here we should have one test ```happyPath``` in which we add some products and then expect thta when we sort them there order is appropriate.

**Comment by EmilGeorgiev** [comment]
LGTM after fix comments

**Pull request merged**